downstairs/src/main.rs

```diff
@@ -587,24 +641,39 @@ async fn proc(ds: &Arc<Downstairs>, mut sock: TcpStream) -> Result<()> {
                  * Negotiate protocol before we get into specifics.
                  */
                 match new_read.transpose()? {
-                    None => return Ok(()),
-                    Some(Message::HereIAm(version)) => {
+                    None => {
+                        println!("upstairs {:?} disconnected", upstairs_uuid.unwrap());
+                        return Ok(());
+                    }
+                    Some(Message::HereIAm(version, uuid)) => {
                         if negotiated {
                             bail!("negotiated already!");
                         }
                         if version != 1 {
                             bail!("expected version 1, got {}", version);
                         }
                         negotiated = true;
+                        upstairs_uuid = Some(uuid);
+                        println!("upstairs {:?} connected", upstairs_uuid.unwrap());
                         fw.send(Message::YesItsMe(1)).await?;
                     }
+                    Some(Message::PromoteToActive(uuid)) => {
+                        // Only allowed to promote or demote self
+                        assert_eq!(upstairs_uuid.unwrap(), uuid);
```

Comment:
Yes - see https://github.com/oxidecomputer/crucible/pull/75#discussion_r723306166

Reply:
The way I figure it is in that situation offset and/or length is coming from Guest input, which is not in our control in the same way UUID is. UUID should also never change unless the Upstairs restarts - it's not per-request.

But after sleeping on it, I also think that this should be an error instead now. I agree we want to avoid panics.

Reply:
Changed to return a UuidMismatch message in 58b5850